cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/publisher/KafkaMetadataChangePublisher.java

```diff
@@ -47,38 +51,38 @@
   private static final Gson GSON = new GsonBuilder()
     .registerTypeAdapter(Id.NamespacedId.class, new NamespacedIdCodec())
     .create();
-  private static final String PUBLISHER_CACHE_KEY = "kafkaPublisher";
   private final KafkaPublisher.Ack ack;
-  // This cache will always only contain a single entry, since we will always use the same key.
-  // However, using it will take care of synchronization issues while creating a publisher.
-  private final LoadingCache<String, KafkaPublisher> publishers;
+  private final Supplier<Producer<Integer, ByteBuffer>> producer;
   private final String topic;
+  private final String brokerList;
 
   @Inject
-  KafkaMetadataChangePublisher(CConfiguration cConf, final KafkaClient kafkaClient) {
+  KafkaMetadataChangePublisher(CConfiguration cConf) {
     this.ack = KafkaPublisher.Ack.FIRE_AND_FORGET;
     this.topic = cConf.get(Constants.Metadata.UPDATES_KAFKA_TOPIC);
-    this.publishers = CacheBuilder.newBuilder().build(new CacheLoader<String, KafkaPublisher>() {
+    this.brokerList = cConf.get(Constants.Metadata.KAFKA_BROKER_LIST);
+    this.producer = Suppliers.memoize(new Supplier<Producer<Integer, ByteBuffer>>() {
       @Override
-      @SuppressWarnings("NullableProblems")
-      public KafkaPublisher load(String key) throws Exception {
-        return kafkaClient.getPublisher(ack, Compression.SNAPPY);
+      public Producer<Integer, ByteBuffer> get() {
+        Properties props = new Properties();
+        props.put("metadata.broker.list", brokerList);
+        props.put("serializer.class", ByteBufferEncoder.class.getName());
+        props.put("key.serializer.class", IntegerEncoder.class.getName());
+        props.put("partitioner.class", IntegerPartitioner.class.getName());
+        props.put("request.required.acks", Integer.toString(ack.getAck()));
+        props.put("compression.codec", Compression.SNAPPY.getCodec());
+        ProducerConfig config = new ProducerConfig(props);
+        return new Producer<>(config);
       }
     });
   }
 
   @Override
   public void publish(MetadataChangeRecord changeRecord) {
-    KafkaPublisher publisher;
-    try {
-      publisher = publishers.get(PUBLISHER_CACHE_KEY);
-    } catch (ExecutionException e) {
-      LOG.warn("Unable to get kafka publisher, will not be able to publish metadata.");
-      return;
-    }
     byte[] changesToPublish = Bytes.toBytes(GSON.toJson(changeRecord));
-    KafkaPublisher.Preparer preparer = publisher.prepare(topic);
-    preparer.add(ByteBuffer.wrap(changesToPublish), changeRecord.getPrevious().getTargetId().hashCode());
-    preparer.send();
+    Object partitionKey = changeRecord.getPrevious().getTargetId().hashCode();
```

Comment:
Since we are doing a hashcode later, this can just be `Object partitionKey = changeRecord.getPrevious().getTargetId()`, right?